### ISSUE TITLE
Steam refused to start after system crash

### ISSUE THREAD

**Issue by h1z1** [ValveSoftware/steam-for-linux#5013]
#### Your system information

* Steam client version (build number or date): 
Build: Jun 1 2017 at  18:41:15

* Distribution (e.g. Ubuntu):  CentOS 7

* Opted into Steam client beta?: No

* Have you checked for system updates?: Yes

#### Please describe your issue in as much detail as possible:

I expected steam to start or at least give some indication as to why it wasn't.  Instead running the client resulted in virtually zero console output and nothing in logs.

#### Steps for reproducing this issue:

1.  System crashed (USB related).  AKA an unclean shutdown.
2.  Restart
3.  Try to start Steam ( in my case a shell script - /bin/steam)
4.  Stare at a blinking cursor

After trying everything including reinstalling Steam entirely, nothing worked until I manually ran the steam runtime

~/.local/share/Steam/ubuntu12_32/steam

That finally bootstrapped the client.   Restarting steam with the usual shell script /bin/steam works as well now.

I tried everything from GDB to strace, letting the client run overnight, nothing of interest.  The client was stuck waiting on a mutex but I had no idea _what_ it was locked on nor why it wouldn't finally timeout.

Maybe it was an unrelated client update on June 1st that fixed it, I don't know.   Hope this helps someone anyway, it was driving me crazy for weeks.

**Comment by kisak-valve** [comment]
Hello @h1z1, if you happen to find a way to consistently reproduce this issue, it would be helpful for troubleshooting it.

**Comment by h1z1** [comment]
Will do.  Short of a corefile which I'm not even sure would be helpful, is there anything specific I can do / include in the future?  All your debugging information appears dependent on steam actually running.

Maybe worth follow up internally?  Given the source to steam is not available it's hard to track where mutex locks are used.  I suspect this had more to do with a lock file or something possibly kept outside of what I'd consider "steam related directories".   Also given it happened early enough that no GUI was drawn at all I'd say it was fairly basic.

I created this bug to track that it happened and maybe help someone else later.

**Comment by h1z1** [comment]
Think I'm hitting this again.   Noticed the client updated,  grabbed Payday2 and restarted steam.  It now refuses to start throwing an error about swrast.  There is no problem with graphics...



```
$ steam
running Steam on centos 7 64-bit
STEAM_RUNTIME is enabled automatically
Installing breakpad exception handler for appid(steam)/version(1496897923)
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast

<< nothing past here >>
```

[GDB output](https://pastebin.com/CgjbYNWL)


What else can I provide?

**Comment by kisak-valve** [comment]
It's odd that steam is trying to use swrast (software rendering from mesa) without trying to use real hardware first. It hints that your 32 bit opengl driver is not setup properly. What video card vendor and driver are you using?

**Comment by h1z1** [comment]
Agreed.  Even more bizzare is it worked fine before restarting steam.

glxinfo:

```
OpenGL vendor string: NVIDIA Corporation
OpenGL renderer string: GeForce GTX 980 Ti/PCIe/SSE2
OpenGL core profile version string: 4.3.0 NVIDIA 375.26
OpenGL core profile shading language version string: 4.30 NVIDIA via Cg compiler
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile
OpenGL core profile extensions:
OpenGL version string: 4.5.0 NVIDIA 375.26
OpenGL shading language version string: 4.50 NVIDIA
OpenGL context flags: (none)
OpenGL profile mask: (none)
OpenGL extensions:

```

Proving 32bit GL works:

```
$ file `which glxspheres`
/bin/glxspheres: ELF 32-bit LSB executable, Intel 80386, version 1 (SYSV), dynamically linked (uses shared libs), for GNU/Linux 2.6.32, BuildID[sha1]=e2b633cc1a10c43ecb192de8fa21af7744d8cc5c, not stripped
$
$ glxspheres
Polygons in scene: 62464
Visual ID of window: 0x27
Context is Direct
OpenGL Renderer: GeForce GTX 980 Ti/PCIe/SSE2
165.010983 frames/sec - 184.152257 Mpixels/sec
164.769010 frames/sec - 183.882215 Mpixels/sec
$ 

```

**Comment by h1z1** [comment]
In the interest of playing _ANY_ steam games this weekend I've run the botstrap manually again. 

 I tried

 - removing ~/.Steam and starting /bin/steam 
 - removing ~/.local/share/Steam  and starting /bin/steam 
 - removing ~/.local/share/Steam/steamapps  and starting /bin/steam 
 - removing ~/.local/share/Steam and manually running ~/.local/share/Steam/ubuntu12_32/steam  this downloaded the client again
- on restart I ran /bin/steam successfully.

There were no other system or environment changes otherwise.  Steam once again starts and without the swrast error.

For shits'n giggles I restored  a backup of~/.local/share/Steam  from before doing anything.  Steam once again fails to start throwing a swrast error.

I'm honestly pretty nervious at this point since it looks like half the client was removed?
$ $ pwd
/home/steam/.local/share
$
$ ls Steam.bak
bootstrap.tar.xz  package        steam_install_agreement.txt  userdata
linux32           steamapps      steam.sh
logs              steamdeps.txt  ubuntu12_32

$ ls Steam
appcache          linux64                  steam_install_agreement.txt
bin               logs                     steam_msg.sh
bin_steamdeps.py  music                    steam.sh
bin_steam.sh      package                  tenfoot
bootstrap.tar.xz  public                   ThirdPartyLegalNotices.css
config            resource                 ThirdPartyLegalNotices.doc
controller_base   servers                  ThirdPartyLegalNotices.html
friends           skins                    ubuntu12_32
graphics          ssfn4620974694217431137  ubuntu12_64
html5app          steam                    userdata
linux32           steamdeps.txt


Going from using steam in this configuration for at least a year, I can say rather confidently "it worked before".
